File: home/actuator.py
import time
from home.logger_config import logger
from home.config import DAILY_ROUTINE_DURATION


class Actuator:
    actuator_count = {}  # Dictionary to keep track of sensor count for each room and sensor type

    def __init__(self, actuator_type, room_name):
        if room_name not in Actuator.actuator_count:
            # Initialize sensor count for the room if not already present
            Actuator.actuator_count[room_name] = {}
        if actuator_type not in Actuator.actuator_count[room_name]:
            # Initialize sensor count for the sensor type if not already present
            Actuator.actuator_count[room_name][
                actuator_type] = 0
        # Increment sensor count for the sensor type in the room
        Actuator.actuator_count[room_name][actuator_type] += 1

        self.id = f"/Actuator/{actuator_type}/{room_name}/{Actuator.actuator_count[room_name][actuator_type]}"
        self.actuator_type = actuator_type
        self.room_name = room_name
        self.status = "off"

    def turn_on(self):
        self.status = "on"
        print(f"{self.id} is turned {self.status}.")
        logger.info(format("\'" + self.id + "\' is turned " + self.status.upper()))

    def turn_off(self):
        self.status = "off"
        print(f"{self.id} is turned {self.status}.")
        logger.info(format("\'" + self.id + "\' is turned " + self.status.upper()))

    def get_status(self):
        print(f"{self.id} current status is {self.status}")
        logger.info(format("\'" + self.id + "\' CURRENT STATUS is: " + self.status.upper()))
        return self.status


class Heater(Actuator):
    def __init__(self, room_name):
        super().__init__("Heater", room_name)
        self.target_temperature = None

    def set_target_temperature(self, target_temperature):
        self.target_temperature = target_temperature
        logger.info(format(f"Set the target temperature of {self.id} to {self.target_temperature}°C."))
        print(f"Set the target temperature of {self.id} to {self.target_temperature}°C.")

    def adjust_temperature(self, current_temperature):
        if self.target_temperature is not None:
            if current_temperature < self.target_temperature:
                self.turn_on()
            else:
                self.turn_off()


class AC(Actuator):
    def __init__(self, room_name):
        super().__init__("AC", room_name)
        self.target_temperature = None

    def set_target_temperature(self, target_temperature):
        self.target_temperature = target_temperature
        logger.info(format(f"Set the target temperature of {self.id} to {self.target_temperature}°C."))
        print(f"Set the target temperature of {self.id} to {self.target_temperature}°C.")

    def adjust_temperature(self, current_temperature):
        if self.target_temperature is not None:
            if current_temperature > self.target_temperature:
                self.turn_on()
            else:
                self.turn_off()


class CoffeeMachine(Actuator):
    def __init__(self, room_name):
        super().__init__("CoffeeMachine", room_name)

    def make_coffee(self, coffee_type):
        if self.status == "on":
            logger.info(format(self.id + "Start making " + coffee_type))
            print(f"{self.id} Start making {coffee_type}")
        elif self.status == "off":
            print(f" {self.id} is OFF now")
            logger.warning(format("\' " + self.id + " \'" + " is OFF now, Need to Turn it On First."))

        else:
            logger.error("There is Some Error with the Coffee Machine" + self.id + ".")
            print("There is Some Error with the Coffee Machine" + self.id + ".")


class Window(Actuator):
    def __init__(self, room_name):
        super().__init__("Window", room_name)


class Door(Actuator):
    def __init__(self, room_name):
        super().__init__("Door", room_name)

    def lock(self):
        logger.info(format("Lock the door" + self.id + "."))
        print(f"Lock the door {self.id}.")

    def unlock(self):
        logger.info(format("Unlock the door" + self.id + "."))
        print(f"Unlock the door {self.id}.")


class Curtain(Actuator):
    def __init__(self, room_name):
        super().__init__("Curtain", room_name)


class CleaningRobot(Actuator):
    def __init__(self, room_name):
        super().__init__("CleaningRobot", room_name)

    def daily_routine(self):
        if self.status == "off":
            logger.warning(format("Cleaning Robot " + self.id + " is OFF now, Need to Turn it ON First"))
            print("Cleaning Robot " + self.id + " is OFF now, Need to Turn it ON First")
        elif self.status == "on":
            logger.info(format("Cleaning Robot " + self.id + " Starts Daily Cleaning Routine"))
            print(f"Cleaning Robot {self.id} Starts Daily Cleaning Routine")
            time.sleep(DAILY_ROUTINE_DURATION)
            logger.info(format(f"{self.id} Finish Daily Cleaning Routine, Will Turn it OFF"))
            print(f"{self.id} Finish Daily Cleaning Routine, Will Turn it OFF")
            self.turn_off()
        else:
            logger.error("There is Some Error with the Cleaning Robot" + self.id + ".")
            print(f"There is Some Error with the Cleaning Robot {self.id}.")


class NotificationSender(Actuator):
    def __init__(self, room_name):
        super().__init__("NotificationSender", room_name)

    def notification_sender(self, message):
        if self.status == "on":
            logger.info(format("Notification Sender " + self.id + " sends message: " + message))
            print(f"Notification Sender {self.id} send message: {message}")
        elif self.status == "off":
            logger.warning(format("Notification Sender " + self.id + " is OFF, Turn it On First."))
            print(f"Notification Sender {self.id} is OFF, please turn it on then send the message again.")
        else:
            logger.error("Fail to send the message. There is some error with the Notification Sender")
            print("Fail to send the message. There is some error with the Notification Sender.")


class MusicPlayer(Actuator):
    def __init__(self, room_name):
        super().__init__("MusicPlayer", room_name)

    def play_music(self, playlist):
        if self.status == "on":
            print(f"{self.id} start playing {playlist}")
            logger.info(format(f"{self.id} start playing {playlist}"))
        elif self.status == "off":
            print(f"Music Player {self.id} is OFF now, Turn it ON First")
            logger.warning(format("Music Player" + self.id + " is OFF now, Turn it ON First"))
        else:
            logger.error("Fail to play " + playlist + ", There is some error with the music player.")
            print("Fail to play " + playlist + ", There is some error with the music player.")


class Light(Actuator):
    def __init__(self, room_name):
        super().__init__("Light", room_name)
        self.brightness_levels = {
            "low": 30,
            "medium": 60,
            "high": 90
        }
        self.brightness_level = 0

    def set_brightness_level(self, level_name):
        if level_name in self.brightness_levels:
            if self.status == "on":
                self.brightness_level = self.brightness_levels[level_name]
                print(f"Set {self.id} light brightness level to {level_name.upper()}")
                logger.info(format("Set the " + self.id + " light brightness level to " + level_name.upper()))
            elif self.status == "off":
                print(f"Light {self.id} is OFF. Please turn it on before setting the brightness level.")
                logger.warning(
                    format("Light " + self.id + " is OFF. Please turn it on before setting the brightness level."))
            else:
                print("There is an error with the Light.")
                logger.error("There is an error with the Light.")
        else:
            print(
                f"Invalid brightness level: {level_name}. Available levels are {list(self.brightness_levels.keys())}.")
            logger.error(format("Invalid brightness level: " + level_name + ". Available levels are " + str(
                list(self.brightness_levels.keys()))))


class SmartTV(Actuator):
    def __init__(self, room_name):
        super().__init__("SmartTV", room_name)

    def play_channel(self, channel_name):
        if self.status == "on":
            logger.info(format("Start playing " + channel_name + " on the " + self.id + " in " + self.room_name))
            print(f"Start playing {channel_name} on the {self.id} in {self.room_name}")
        elif self.status == "off":
            logger.warning(format("\'" + self.id + "\' is OFF now, Need to Turn it ON First."))
            print(f"\'{self.id}\' is OFF now, Need to Turn it ON First.")

        else:
            logger.error("Fail to play " + channel_name + ", There is some error with the TV.")
            print("Fail to play " + channel_name + ", There is some error with the TV.")


class SmartSocket(Actuator):
    def __init__(self, room_name):
        super().__init__("SmartSocket", room_name)


class Humidifier(Actuator):
    def __init__(self, room_name):
        super().__init__("Humidifier", room_name)

    def increase_humidity(self):
        logger.info(f"{self.id} Increasing humidity in {self.room_name}")
        print(f"{self.id} Increasing humidity in {self.room_name}")

    def decrease_humidity(self):
        logger.info(f"{self.id} Decreasing humidity in {self.room_name}")
        print(f"{self.id} Decreasing humidity in {self.room_name}")

File: experiments_js/home/actuator.js
const { logger } = require('./logger_config');
const {DAILY_ROUTINE_DURATION} = require("./config");

class Actuator {
    static actuatorCount = {}; // Track actuator count for each room and actuator type

    constructor(actuatorType, roomName) {
        if (!Actuator.actuatorCount[roomName]) {
            Actuator.actuatorCount[roomName] = {}; // Initialize room if not already present
        }
        if (!Actuator.actuatorCount[roomName][actuatorType]) {
            Actuator.actuatorCount[roomName][actuatorType] = 0; // Initialize actuator type count
        }

        Actuator.actuatorCount[roomName][actuatorType]++; // Increment actuator count

        this.id = `/Actuator/${actuatorType}/${roomName}/${Actuator.actuatorCount[roomName][actuatorType]}`;
        this.actuatorType = actuatorType;
        this.roomName = roomName;
        this.status = "off";
    }

    turnOn() {
        this.status = "on";
        console.log(`${this.id} is turned ${this.status}.`);
        logger.info(`${this.id} is turned ON.`);
    }

    turnOff() {
        this.status = "off";
        console.log(`${this.id} is turned ${this.status}.`);
        logger.info(`${this.id} is turned OFF.`);
    }

    getStatus() {
        console.log(`${this.id} current status is: ${this.status}`);
        logger.info(`${this.id} CURRENT STATUS is: ${this.status.toUpperCase()}`);
        return this.status;
    }
}

class Heater extends Actuator {
    constructor(roomName) {
        super("Heater", roomName);
        this.targetTemperature = null;
    }

    setTargetTemperature(targetTemperature) {
        this.targetTemperature = targetTemperature;
        console.log(`Set the target temperature of ${this.id} to ${this.targetTemperature}°C.`);
        logger.info(`Set the target temperature of ${this.id} to ${this.targetTemperature}°C.`);


    }

    adjustTemperature(currentTemperature) {
        if (this.targetTemperature && currentTemperature < this.targetTemperature) {
            this.turnOn();
        } else {
            this.turnOff();
        }
    }
}

class AC extends Actuator {
    constructor(roomName) {
        super("AC", roomName);
        this.targetTemperature = null;
    }

    setTargetTemperature(targetTemperature) {
        this.targetTemperature = targetTemperature;
        console.log(`Set the target temperature of ${this.id} to ${this.targetTemperature}°C.`);
        logger.info(`Set the target temperature of ${this.id} to ${this.targetTemperature}°C.`);
    }

    adjustTemperature(currentTemperature) {
        if (this.targetTemperature && currentTemperature > this.targetTemperature) {
            this.turnOn();
        } else {
            this.turnOff();
        }
    }
}

class CoffeeMachine extends Actuator {
    constructor(roomName) {
        super("CoffeeMachine", roomName);
    }

    makeCoffee(coffeeType) {
        if (this.status === "on") {
            console.log(`${this.id} Start making ${coffeeType}`);
            logger.info(`${this.id} Start making ${coffeeType}`);
        } else if (this.status === "off") {
            console.log(`${this.id} is OFF now.`);
            logger.warning(`${this.id} is OFF now, Need to turn it ON First.`);
        } else {
            logger.error(`There is some error with the Coffee Machine ${this.id}.`);
            console.log(`There is some error with the Coffee Machine ${this.id}.`);
        }
    }
}


class Window extends Actuator {
    constructor(roomName) {
        super("Window", roomName);
    }
}

class Door extends Actuator {
    constructor(roomName) {
        super("Door", roomName);
    }

    lock() {
        console.log(`Lock the door ${this.id}.`);
        logger.info(`Lock the door ${this.id}.`);
    }

    unlock() {
        console.log(`Unlock the door ${this.id}.`);
        logger.info(`Unlock the door ${this.id}.`);
    }
}


class Curtain extends Actuator {
    constructor(roomName) {
        super("Curtain", roomName);
    }
}

class CleaningRobot extends Actuator {
    constructor(roomName) {
        super("CleaningRobot", roomName);
    }

    dailyRoutine() {
        if (this.status === "off") {
            logger.warning(`Cleaning Robot ${this.id} is OFF, turn it ON First.`);
            console.log(`Cleaning Robot ${this.id} is OFF now, Need to Turn it ON First.`)
        }
        else if (this.status === "on") {
            logger.info(`Cleaning Robot ${this.id} Starts Daily Cleaning Routine.`);
            console.log(`Cleaning Robot ${this.id} Starts Daily Cleaning Routine.`);

            setTimeout(() => {
                console.log(`${this.id} Finish Daily Cleaning Routine, Will Turn it OFF.`);
                logger.info(`${this.id} Finish Daily Cleaning Routine, Will Turn it OFF.`);
                this.turnOff();
            }, DAILY_ROUTINE_DURATION * 1000);
        } else {
            console.log(`There is Some Error with the Cleaning Robot ${this.id}.`);
            logger.error(`There is Some Error with the Cleaning Robot ${this.id}.`);
        }
    }
}

class NotificationSender extends Actuator {
    constructor(roomName) {
        super("NotificationSender", roomName);
    }

    sendNotification(message) {
        if (this.status === "on") {
            console.log(`Notification from ${this.id} send message: ${message}`);
            logger.info(`Notification from ${this.id} send message: ${message}`)
        } else if (this.status === "off") {
            console.log(`${this.id} is OFF, turn it on first.`);
            logger.warning(`Notification Sender ${this.id} is OFF, Turn it On First.`);
        } else{
            logger.error("Fail to send the message. There is some error with the Notification Sender")
            console.log("Fail to send the message. There is some error with the Notification Sender.")
        }
    }
}

class MusicPlayer extends Actuator {
    constructor(roomName) {
        super("MusicPlayer", roomName);
    }

    playMusic(playlist) {
        if (this.status === "on") {
            console.log(`${this.id} start playing ${playlist}`);
            logger.info(`${this.id} start playing ${playlist}`)
        } else if(this.status === "off") {
            console.log(`${this.id} is OFF now, Turn it ON First.`);
            logger.warning(`${this.id} is OFF now, Turn it ON First.`);
        } else{
            logger.error(`Fail to play ${playlist}, There is some error with the music player.`)
            console.log(`Fail to play ${playlist}, There is some error with the music player.`)
        }
    }
}

class Light extends Actuator {
    constructor(roomName) {
        super("Light", roomName);
        this.brightnessLevels = { low: 30, medium: 60, high: 90 };
        this.brightnessLevel = 0;
    }

    setBrightness(level) {
        if (this.brightnessLevels[level] !== undefined) {
            if (this.status === "on") {
                this.brightnessLevel = this.brightnessLevels[level];
                console.log(`Set ${this.id} light brightness level to ${level.toUpperCase()}`);
                logger.info(`Set the ${this.id} light brightness level to ${level.toUpperCase()}`);
            }else if (this.status === "off"){
                console.log(`Light ${this.id} is OFF. Please turn it on before setting the brightness level.`);
                logger.warning(`Light ${this.id} is OFF. Please turn it on before setting the brightness level.`);
            }else {
                console.log(`There is an error with the Light ${this.id}`);
                logger.error(`There is an error with the Light ${this.id}`);
            }
        } else {
            console.log(`Invalid brightness level: ${level}`);
            logger.error(`Invalid brightness level: ${level}`);
        }
    }
}

class SmartTV extends Actuator {
    constructor(roomName) {
        super("SmartTV", roomName);
    }

    playChannel(channelName) {
        if (this.status === "on") {
            console.log(`Start playing ${channelName} on ${this.id} in ${this.roomName}`);
            logger.info(`Start playing ${channelName} on ${this.id} in ${this.roomName}`);
        } else if(this.status === "off") {
          logger.warning(`${this.id} is OFF now, Need to Turn it ON First.`);
            console.log(`${this.id} is OFF now, Need to Turn it ON First.`);
        } else {
            console.log(`Fail to play ${channelName}, There is some error with the TV.`);
            logger.error(`Fail to play ${channelName}, There is some error with the TV.`);
        }
    }
}

class SmartSocket extends Actuator {
    constructor(roomName) {
        super("SmartSocket", roomName);
    }
}

class Humidifier extends Actuator {
    constructor(roomName) {
        super("Humidifier", roomName);
    }

    increaseHumidity() {
        logger.info(`${this.id} Increasing humidity in ${this.roomName}.`);
        console.log(`${this.id} Increasing humidity in ${this.roomName}.`);
    }

    decreaseHumidity() {
        console.log(`${this.id} Decreasing humidity in ${this.roomName}.`);
        logger.info(`${this.id} Decreasing humidity in ${this.roomName}.`);
    }
}



module.exports = { Actuator, Heater, AC, CoffeeMachine, Window, Door, Curtain, CleaningRobot, NotificationSender, MusicPlayer, Light, SmartTV, SmartSocket, Humidifier};